Added Warning Banner for local kdbx files

stalkopat opened the pull request:
#159 

Added a Warning Banner widget [BackupBanner] that displays the warning.
Dismiss persitance is achieved inside app_data with a BuildList containing file uuids.

Isolated a new widget [SaveFileAs] from ManageFile wich saves a given file either locally or inside a provided CloudStorageProvider. For each CloudStorageProvider a new SaveFileAs widget is created to increase reusability.

I decided against using a BottomModalSheet because currently there are 4 Cloud Storages, 5 including Local Storage,
incase more Cloud Storage Providers get added later on accessibility could get impacted.

For the backup options inside the Banner a showMenu with all the SaveFileAs's was created.

Added some Localizations for saving and backup messages.

Feedback very much  appreciated! 😄 

Edit: Fixed the branches and my git user. Didn't have it configured properly, sorry about that

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/authpass/authpass?pullRequest=167) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/authpass/authpass?pullRequest=167) before we can accept your contribution.<br/><hr/>**Thomas Salzgeber** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/authpass/authpass?pullRequest=167) it.</sub>

hpoul reviewed (changes requested):
Hey, all in all looks pretty good 👍️ thanks so much! A couple of things I would suggest:

* there seems to be a merge conflict in the `password_list.dart` (a `)` missing 😅️)
* It would be nice if we could share more code between the manage_file and the password list - So i think a generic method for a 'shareAs' dialog which can be used in both places. either just a function which more or less wraps `showMenu` or a widget which is used as either as `showDialog` or `showModalBottomSheet`.. either way would be good i guess.
* There seem to be some double listeners, and some weird interaction that the child `ListTile` has an action as well as the parent pop up menu. so you have two hover effects, and when you click outside the inner hover effect, the menu just disappears without doing anything:

<img width="229" alt="Screen Shot 2020-10-01 at 20 09 12" src="https://user-images.githubusercontent.com/313066/94846859-1bfe9380-0422-11eb-997f-efe06d72249b.png">

<img width="159" alt="Screen Shot 2020-10-01 at 20 09 26" src="https://user-images.githubusercontent.com/313066/94846877-2456ce80-0422-11eb-9219-9a146c3221e4.png">

* Regarding wording: I think i wouldn't call the action 'Backup' because this could be interpreted as it is being a one time backup, and no long term switch of storage backend. What do you think about something like: "Your passwords in XXX are only saved locally! We recommend using a cloud storage service to keep your data save." [ Dismiss ] [ Save As ]

(I think dismiss buttons should go to the left according to the [material guidelines](https://material.io/components/dialogs#actions) - although i'm not sure i've been consistently following this myself) --- internally it can still be called `BackupBanner` 👍️
* We should at most display one banner at a time. (again a [material guidelines](https://material.io/components/banners#usage), but would also be a bit annoying :-) ) so just don't `map` through all files, but get the first one. - also what would you think about only showing the banner when there are more than like.. 3 passwords in the database? This way new users won't see it the first time they create their new password file?

let me know if you agree with those points, great work 👍️ thanks a ton!

stalkopat commented:
Sorry about the merge conflict, probably deleted one line when merging with github UI.

The PopupMenuItem has a hardcoded padding of 16dp wich causes the weird UI.
![image](https://user-images.githubusercontent.com/25117613/94899083-f2ca1b80-0492-11eb-9d78-f7f489f8df05.png)
A Problem i faced was the fact that either you have PopupMenuItems containing the ListTile and handle all the interaction outside of the ListTile and together with the PopupMenuButton / showMenu wich severly limits reusability or, 
as i did, emulate the PopupMenuItem actions by giving the ListTile an onTap handler wich pops the current route, which is what PopupMenuItem is doing under the hood. I felt like the second approach had more flexibility since a ListTile can appear anywhere in the App.

But editing Flutter framework files is obviously a deal breaker, i'll see what i can do, i guess implementing PopupMenuEntry would work, i'll figure something out for showing dialogs while keeping reusability.

I feel like Save As implies that the format is being changed, from say kdbx to some other Password file format. How about "Upload to" or "Move to"

I agree with the Button placement and only displaying one at a time.

Having a treshhold for displaying the banner is a good point, do you have metrics on how many database's a user has on avergae and how many passwords per database are stored?

hpoul commented:
no problem @ conflict.

ad naming: from these two i'd prefer `Move to` .. maybe `Save to` ? :-) I actually thought `Save as ...` would be the standard language, like in desktop apps:

<img width="375" alt="Screen Shot 2020-10-02 at 10 19 21" src="https://user-images.githubusercontent.com/313066/94902787-0a0c0780-0499-11eb-9af6-2b7b1c893101.png">

(Numbers also only supports one format in `save as`, all other formats would be `export to`). But i'd be fine with `Save to` ..  we could also do a `Move to` but, i think then we should delete the old file :-) )

To be honest, I think at this point I'd prefer simply using a modal bottom sheet, instead of a menu altogether. This might be a bit weird on desktop, but on mobile apps drop down menus are weird (exception being the overflow menu). in a modal bottom sheet we could simply use the `onTap` of the `ListItem` - I don't think there would be much downside? and it makes the code easier, since you don't have to get the tap location for placing the menu 😅️

Nope, I'd expect 99% having only one, I've even got people answering in the survey that they wanted an easy way to switch between files, before i told them that they could just open both at the same time :-D but anyway.. it would be awkward having two banners on top of each other..

Re. number of passwords: My thought was, unless you've got >=3 passwords, it's probably not too bad to lose them :-)

stalkopat commented:
Created a custom PopupMenuButton esque Widget that depending on the platform displays either a Popup menu or a BottomModal sheet. Creating a custom PopupMenuEntry worked fine, UI glitches are gone and its usable inside a ModalBottomSheet. I also feel like the code inside password_list and manage_file is quiet clean now: 
![image](https://user-images.githubusercontent.com/25117613/94923785-b6131a00-04bc-11eb-9072-f60cbb3752d6.png)

hpoul commented on authpass/lib/ui/widgets/utils/save_file_helper.dart:
```diff
@@ -0,0 +1,138 @@
+    } else {
+      return PopupMenuButton<KdbxOpenedFile>(
+        child: widget.child ?? widget.child,
+        onSelected: (_) => _,
```
this noop kind of bothers me, what happens when we remove the parameter altogether?

stalkopat replied:
missed that during refactoring, that param can go...

hpoul reviewed (changes requested):
I like it, works great. 👍️ just a few minor code-level things.. Sorry if I'm a bit pedantic.. I don't have much experience reviewing (non-coworkers) either 😅️

* I've left a few inline comments/suggestions, please take a look
* I think `save_file.dart` and `save_file_helper.dart` should live together, in the same directory.. and we might need to come up with more descriptive names like `save_as_menu.dart`, `save_as_button.dart` - but I like the separation, I'm usually too lazy 😅️
* there are a few unused imports
* some personal preferences regarding code style - if you deliberately did it another way that's fine with me as well
  * Generally I prefer having all parameter lists end with a trailing `,` to have it multiline (when it's longer than 1 line)
  * Once there are multiple parameters, I tend to make them all named, instead of 2 positional, and all other named.
  * In newer code i've mostly switched to `context.read` and `context.watch` extension methods vs. `Provider.of` (`read `is for callback, `watch` for build methods)
  * try not to implement things we don't use 😅️ (`boolean cloud` switch, `onClose` callback)

thanks again for all your work.

stalkopat commented on authpass/lib/ui/widgets/utils/save_file_helper.dart:
```diff
@@ -0,0 +1,138 @@
+  _SaveFileAsDialogButtonState createState() => _SaveFileAsDialogButtonState();
+}
+
+class _SaveFileAsDialogButtonState extends State<SaveFileAsDialogButton> {
```
SaveFileAsDialogButton can indeed be changed to stateless. SaveFileAsMenuItem on the other hand can't because it extends PopupMenuEntry...

stalkopat commented:
Fixed a small bug where a dismissed message still triggered a banner build wich resulted in Bad Element.

Changed all params to named and '@required'

I wasn't aware of context.read / context.watch, way cleaner than Provider.of<T>(context, listen:false)! Thank you! 😄 

Also moved every widget into its own file under widgets/savefile/...

Thanks for the feedback! I appreciate it!

hpoul commented:
@stalkopat thanks! 👍️ looks good, I just did a small refactor and removed support for drop down menu, because it added a bit of unnecessary complexity.. i think it's not too bad on desktop this way either. (And currently it crashed because it used context.watch instead of context.read - probably because `getItems` was called during `build` for bottom sheets, but as a response to a user action for menus).

Maybe you can help me test this refactoring 😅️ somehow I discovered a few additional bugs which were also there before.. (like saving to google drive 🤦️)

stalkopat commented:
Seems to work for me.

I kinda disagree with the drop down / bottom sheet thing (as per https://material.io/components/sheets-bottom#modal-bottom-sheet) but thats ultimately a stylistic choice. 

I didn't build for Desktop and just tested each individually with removing the If tags. Replacing Provider with context is probably where the bug sneaked in (context.read gives exceptions outside of callbacks) and on mobile it seemed fine. My bad 😅️.

Seems good to me! 👍️

hpoul commented:
> I kinda disagree with the drop down / bottom sheet thing (as per https://material.io/components/sheets-bottom#modal-bottom-sheet) but thats ultimately a stylistic choice.

funny, I didn't see the "Modal bottom sheets are used in mobile apps only" before 🤔️ especially since [non-modal bottom sheets seem to be just fine](https://lh3.googleusercontent.com/Jrlu5zjn0b5TTxZ92S6woxZOwr8s-9UV2up31ACM_0ch1pgWYq81dNj5NgEz4or2uHQ49kXsvrFk1xvfh6dazL9TocUYwhKOK1cA=w1064-v0) 😅️ 

anyway, I would stick with it for now, seems like a separate issue ;) (and then we should probably decide more on screen size, not platform, as tablets/iPad probably should also not have a modal bottom sheet) 🤷️

I'll go ahead and merge it and push to beta.. let's see if someone stumbles on it ;-)

great work 👍️ thanks!
